keps/sig-node/0035-20190130-topology-manager.md

```diff
@@ -172,39 +173,89 @@ Kubelet components.
 
 The Topology Manager component will be disabled behind a feature gate until
 graduation from alpha to beta.
 
 #### Computing Preferred Affinity
 
-A topology hint indicates a preference for some well-known local resources.
-Initially, the only supported reference resource is a mask of CPU socket IDs.
-After collecting hints from all providers, the Topology Manager chooses some
-mask that is present in all lists. Here is a sketch:
-
-1. Apply a partial order on each list: number of bits set in the
-   mask, ascending. This biases the result to be more precise if
-   possible.
-1. Iterate over the permutations of preference lists and compute
-   bitwise-and over the masks in each permutation.
-1. Store the first non-empty result and break out early.
-1. If no non-empty result exists, return an error.
+After collecting hints from all providers, the Topology Manager delegates the 
+affinity calcuation to the selected Topology Manager Policy.
+
+**best-effort & restricted**:
+1. Loops through the list of hint providers and saves an accumulated list of 
+   the hints returns by each hint provider.
+2. Iterate through all permutations of hints accumulated in Step 1. The hint affinites are merged to a single hint
+   by performing a bitwise AND. The preferred field on the merged hint is set to false if any of the hints in the 
+   permutation returned a false preferred.
+3. The hint with the narrowest preferred affinity is returned.     
+4. If no hint with at least one NUMA Node set is found, return a default hint which is a hint
+   with all NUMA Nodes set and preferred set to false.
+
+**single-numa-node**:
+1. Loops through the list of hint providers and saves an accumulated list of 
+   the hints returns by each hint provider.
+2. Iterate through all permutations of hints accumulated in Step 1. The hint affinites are merged to a single hint
+   by performing a bitwise OR. The preferred field on the merged hint is set to false if the result of the bitwise OR
+   results in a cross NUMA Node allocation, if all resources can be aligned on a single NUMA Node preferred is set to true.
+3. The hint with the narrowest preferred affinity is returned.     
+4. If no hint with at least one NUMA Node set is found, return a default hint which is a hint
+   with all NUMA Nodes set and preferred set to false.
 
 The behavior when a match does not exist is configurable, as described
-above.
+above in the Policies section.
 
 #### New Interfaces
 
 ```go
+package socketmask
+
+// SocketMask interface allows hint providers to create SocketMasks for TopologyHints
+type SocketMask interface {
+	Add(sockets ...int) error
+	Remove(sockets ...int) error
+	And(masks ...SocketMask)
+	Or(masks ...SocketMask)
+	Clear()
+	Fill()
+	IsEqual(mask SocketMask) bool
+	IsEmpty() bool
+	IsSet(socket int) bool
+	IsNarrowerThan(mask SocketMask) bool
+	String() string
+	Count() int
+	GetSockets() []int
+}
+
+type socketMask uint64
```

Comment:
Why did we get rid of presenting;
```
func NewSocketMask(sockets ...int) (SocketMask, error) { ... }
```
in favor of presenting:
```
type socketMask uint64
```

The former is what is usable by other components. The latter is internal to the abstraction.

Reply:
Good point, changed